File: 프로그래머스/2/42586. 기능개발/기능개발.cpp
#include <string>
#include <vector>

using namespace std;

vector<int> solution(vector<int> progresses, vector<int> speeds) {
    vector<int> answer;
    
    int plen=progresses.size();
    int curprogress=0;
    while(1){
        if(curprogress>=plen){
            break;
        }
        
        for(int i=curprogress;i<plen;i++){
            int& progress=progresses.at(i);
            int speed=speeds.at(i);
            
            progress+=speed;
        }
        
        if(progresses.at(curprogress)>=100){
            int count=0;
            for(int i=curprogress;i<plen;i++){
                int progress=progresses.at(i);
                if(progress>=100){
                    count++;
                }
                else{
                    break;
                }                
            }
            answer.push_back(count);
            curprogress+=count;
        }
    }
    
    return answer;
}

File: 프로그래머스/2/42586. 기능개발/기능개발.java
import java.util.*;
class Solution {
    public class Node{
        public int cur;
        public int speed;
        
        public Node(int _cur, int _speed){
            cur=_cur; speed=_speed;
        }
    }
    
    public int[] solution(int[] progresses, int[] speeds) {
        int len=progresses.length;
        int[] answer = {};
        
        Deque<Node> dq=new ArrayDeque<>();
        for(int i=0;i<len;i++){
            Node n=new Node(progresses[i],speeds[i]);
            dq.addLast(n);
        }
        
        List<Integer> result=new ArrayList<>();
        
        while(!dq.isEmpty()){
            
            
            int dlen=dq.size();
            for(int i=0;i<dlen;i++){
                Node cur=dq.pollFirst();
                cur.cur+=cur.speed;
                dq.addLast(cur);
            }
            
            int count=0;
            while(!dq.isEmpty()){
                if(dq.peekFirst().cur>=100){
                    ++count;
                    dq.pollFirst();
                }
                else{
                    break;
                }
            }
            
            if(count>=1){
                result.add(count);
            }
        }
        
        int rlen=result.size();
        answer=new int[rlen];
        for(int i=0;i<rlen;i++){
            answer[i]=result.get(i);
        }
        
        return answer;
    }
}